Improve the documentation on Operations (#477)
wasp-lang/wasp: web/docs/language/features.md

## Queries and Actions (aka Operations)

In Wasp, the client and the server interact with each other through Operations.
Wasp currently supports two kinds of Operations: **Queries** and **Actions**.

### Query

> Queries are resource-fetching functions that don't modify the state. Each Query should fetch, process, and return a particular resource.

Reviewer: I realize the word "resource" was already in the old text, but that doesn't mean we shouldn't review it. Resource is a bit overloaded on the web. It may be easily confused with the notion of an HTTP resource. Why even use the word? What is a "particular resource"? The word seems rather fuzzy in its definition; we're really talking about query results and they're not particular at all but could have overlaps.

"Queries are used to fetch data from the server. Queries, unlike actions, do not modify the server's state."

Author: I agree, your suggestion is much clearer.

Reviewer 2: nice, I like it!

Reviewer 3: > "Queries are used to fetch data from the server. Queries, unlike actions, do not modify the server's state."

This might a bit too specific though -> they don't necessarily have to fetch data from the server, they could fetch data from some external service, or maybe generate data themselves. That said, they are executing on the server (for now), so it is in a way correct that the data is coming from the server, even if originating from somewhere else or constructed on the fly. Okkkkk

Queries are implemented in NodeJS and executed within the server's context.

> However, Wasp will generate the code that lets you call the Query from any point in your code (i.e., client, server) using the same interface.

Reviewer: Starting a sentence with "However," is not really very idiomatic English. Prefer "But" but I'd just leave it out here as what's the contradiction really? In general it's a good idea to avoid the word "will" in sentences; you can often do so and the sentence becomes simpler

"Wasp generates code that lets you call the Query from anywhere in your code (client or server) using the same interface."

(I also got rid of the `i.e.` as client and server are the only two example where you can do it; we have no idea yet whether the upcoming jobs let you use queries).

Author: I'll drop the however. Don't know why it's there, could be a leftover from one of my previous iterations of the text.

I can definitely remove the `i.e.` for cosmetic purposes ("client or server" looks much nicer than "i.e., client, server"), but I think [the two options are equivalent](https://www.merriam-webster.com/words-at-play/ie-vs-eg-abbreviation-meaning-usage-difference) from a semantic standpoint.

Since both arguments are positional, you can name the parameters however you want, but we'll stick with `args` and `context`:
1. `args`:  An object containing all the arguments (i.e., payload) **passed to the Query by the caller** (e.g., filtering conditions).
Take a look at [the examples of usage](#using-the-query) to see how to pass this object to the Query.
3. `context`: An additional context object **injected into the Query by Wasp**. This object contains user session information, as well as information about entities. The examples here won't use the context for simplicity purposes. You can read more about it in the [section about using entities in queries](#using-entities-in-queries).

Here's an example of two simple Queries:
```js title="ext/queries.js"
// our "database"
const tasks = [
  { id: 1, description: "Buy some eggs", isDone: true },
  { id: 2, description: "Make an omelette", isDone: false },
  { id: 3, description: "Eat breakfast", isDone: false }
]

// You don't need to use the arguments if you don't need them
export const getAllTasks = async () => {
  return tasks;
}

// The 'args' object is something sent by the caller (most often from the client)

> export const getFilteredTasks = async (args) => {

Author: Hm, this wouldn't really be correct. We're passing a filtering parameter (`{ isDone: true}`), that's why it returns completed tasks in our use case, but the `isDone` filter could have been anything.

> export const getFilteredTasks = async (args) => {

Reviewer 2: ah yes sorry, get it now!

}
```

#### Declaring a Query in Wasp
After implementing your Queries in NodeJS, all that's left to do before using them is tell Wasp about it!
You can easily do this with the `query` declaration, which supports the following fields:
- `fn: ExtImport` (required) - The import statement of the Query's NodeJs implementation.
- `entities: [Entity]` (optional) - A list of entities you wish to use inside your Query.
We'll leave this option aside for now. You can read more about it [here](#using-entities-in-queries).

Wasp Queries and their implementations don't need to (but can) have the same name, so we will keep the names different to avoid confusion.
With that in mind, this is how you might declare the Queries that use the implementations from the previous step:
```c title="main.wasp"
// ...

query fetchAllTasks {
  fn: import { getAllTasks } from "@ext/queries.js"
}

query fetchFilteredTasks {
  fn: import { getFilteredTasks } from "@ext/queries.js"
}
```

After declaring a NodeJS function as a Wasp Query, two crucial things happen:
- Wasp **generates a client-side JavaScript function** that shares its name with the Query (e.g., `fetchFilteredTasks`).

> This function can take a single argument - an object containing any serializable data you wish to use inside the Query.

Author: It definitely does, but I wanted to emphasize that the argument is optional.

Generating two such functions ensures a uniform calling interface across the entire app (both backend and frontend).

#### Using the Query
To use the Query, you can import it from `@wasp` and call it directly. As mentioned, the usage is the same regardless of whether you're on the server or the client:
```javascript
import fetchAllTasks from '@wasp/queries/fetchAllTasks.js'
import fetchFilteredTasks from '@wasp/queries/fetchFilteredTasks.js'

// ...

const allTasks = await fetchAllTasks();
const doneTasks = await fetchFilteredTasks({isDone: true})
```

**NOTE**: Wasp will not stop you from importing a Query's NodeJS implementation from `./queries.js` and calling it directly. However, we advise against this, as you'll lose all the useful features a Wasp Query provides (e.g., entity injection).

#### The `useQuery` hook
When using Queries on the client, you can make them reactive with the help of the `useQuery` hook.
This hook comes bundled with Wasp and is a thin wrapper around the `useQuery` hook from [_react-query_](https://github.com/tannerlinsley/react-query).

> Wasp's `useQuery` hook takes three arguments:

Reviewer 2: ```suggestion
Wasp's `useQuery` hook accepts three arguments:
```
To make it sound more fancy :D But no need really to change it, I'm just writing things as they come to my mind :D

  })
}
```

The object `context.entities.Task` exposes `prisma.task` from [Prisma's CRUD API](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/crud).

### Action

Actions are very similar to Queries. So similar, in fact, we will only list the differences:
1. They can (and most often should) modify the server's state, while Queries are only allowed to read it.
2. Since Actions don't need to be reactive, Wasp doesn't provide a React hook for them (like `useQuery` for Queries) - you just call them directly.
3. `action` declarations in Wasp are mostly identical to `query` declarations. The only difference is in the declaration's name.

Here's an implementation of a simple Action:

```js title=actions.js
export const sayHi = async () => {
  console.log('The client said Hi!')
}
```
Its corresponding declaration in Wasp:

```c title="main.wasp"
// ...

action sayHi {
  fn: import { sayHi } from "@ext/actions.js"
}
```
And an example of how to import and call the declared Action:

```js
import sayHi from '@wasp/actions/sayHi'

// ...

sayHi()
```

More differences and Action/Query specific features will come in future versions of Wasp.

### Cache Invalidation

> One of the trickiest parts of managing a web app's state is making sure the data shown by the queries is up to date.

Reviewer 2: ```suggestion
One of the trickiest parts of managing a web app's state is making sure the data returned by the queries is up to date.
```
Maybe a bit more technically correct? Since e.g. React does the showing part.

> Since Wasp uses _react-query_ for Query management, we must make sure to invalidate its caches whenever they become stale.

Reviewer 2: ```suggestion
Since Wasp uses _react-query_ for Query management, we must make sure to invalidate queries (more specifically, their cached results managed by react-query) whenever they become stale.
```
Would this be ok? I just found it a bit abrupt to switch to caches from queries so quickly (although it might have been explained above already, sorry if I missed it).

It's possible to invalidate the caches manually through several mechanisms _react-query_ provides (e.g., refetch, direct invalidation).
However, since manual cache invalidation quickly becomes complex and error-prone, Wasp offers a quicker and a more effective solution to get you started: **automatic Entity-based Query cache invalidation**.
Because Actions can (and most often do) modify the state while Queries read it, Wasp invalidates a Query's cache whenever an Action that uses the same Entity is executed.

> For example, let's assume that Action `A1` and Query `Q1` both use Entity `E1`. If `A1` is executed, `Q1`'s cached result may no longer be up-to-date.

Reviewer 2: ```suggestion
For example, let's assume that Action `createTask` and Query `getTasks` both use Entity `Task`. If `createTask` is executed, `getTasks`'s cached result may no longer be up-to-date.
```
I always find it easier to have "real" examples :)

> Wasp will therefore invalidate it, making `Q1` up to date again.

Reviewer 2: ```suggestion
Wasp will therefore invalidate it, making `getTasks` refetch data from the server, making it up to date again.
```

In practice, this means that Wasp keeps the queries "fresh" without requiring you to think about cache invalidation.

> On the other hand, this kind of automatic cache invalidation can become wasteful (some updates might not be necessary) and will not only work for Entities. If that's an issue, you can use the mechanisms provided by _react-query_ for now, and expect more direct support in Wasp for handling those use cases in a nice, elegant way.

Reviewer 2: ```suggestion
On the other hand, this kind of automatic cache invalidation can become wasteful (some updates might not be necessary) and will only work for Entities. If that's an issue, you can use the mechanisms provided by _react-query_ for now, and expect more direct support in Wasp for handling those use cases in a nice, elegant way.
```
Did you mean it like this? Although still not completely sure what you meant to say here?

Author: It's a mistake, nice catch. Leftover "not" from one of the previous iterations.

export const getAllTasks = async () => {
  return tasks;
}

// The 'args' object is something sent by the caller (most often from the client) 
export const getFilteredTasks = async (args) => {
  const { isDone } = args;
  return tasks.filter(task => task.isDone === isDone)
}
```

#### Declaring a Query in Wasp 
After implementing your Queries in NodeJS, all that's left to do before using them is tell Wasp about it!
You can easily do this with the `query` declaration, which supports the following fields:
- `fn: ExtImport` (required) - The import statement of the Query's NodeJs implementation 
- `entities: [Entity]` (optional) - A list of entities you wish to use inside your Query.
We'll leave this option aside for now. You can read more about it [here](#using-entities-in-queries).

Wasp Queries and their implementations don't need to (but can) have the same name, so we will keep the names different to avoid confusion.
With that in mind, this is how you might declare the Queries that use the implementations from the previous step:
```c title="main.wasp"
// ...

query fetchAllTasks {
  fn: import { getAllTasks } from "@ext/queries.js"
}

query fetchFilteredTasks {

>   fn: import { getFilteredTasks } from "@ext/queries.js"

Author: I changed the `query` names here to emphasize the difference between a Wasp query and its NodeJS implementation.

Reviewer 3: Aha, got it!
It is good to demonstrate this, but it sucks a bit that names are not consistent, as there is no good reason to call some of them fetch and some of them get. But ok, I guess it is ok for the purposes of an example!

Author: Yes, the names are definitely arbitrary.

I couldn't come up with any other way to make it clear that what we're using are the things declared in Wasp. I also use these terms later in the text when explaining the differences between the query (`fetchFilteredTasks`) and its implementation (`getFilteredTasks`). We were using the term query interchangeably for these two concepts, and I thought it could cause confusion.

I can change the names to be the same again if you want. However, an even better option would be to keep them different but meaningful (ideas welcome :)).

Reviewer 2: Great exchange here, this definitely reflects what I also experienced while developing Wasp apps. Usually, it is the same name, but we currently can't avoid stating it twice. As an only additional thing I can think of we could do right now is add a comment above query - e.g. "this doesn't have to be the same name, although in practice it often will be."

Author: We already mention that in the text, but I can add a comment too, no problem.

Wasp's `useQuery` hook takes three arguments:
- `queryFn` (required): A Wasp query declared in the previous step or, in other words, the client-side query function generated by Wasp based on a `query` declaration.
- `queryFnArgs` (optional): The arguments object (payload) you wish to pass into the query. The query's NodeJS implementation will receive this object as its first positional argument.
- `config` (optional): A _react-query_ `config` object.

Wasp's `useQuery` hook behaves mostly the same as [_react-query_'s `useQuery` hook](https://react-query.tanstack.com/docs/api#usequery), the only difference being in not having to supply the key. Wasp handles the caching automatically under the hood.

Here's an example of calling the Queries using the `useQuery` hook:
```jsx
import React from 'react'
import { useQuery } from '@wasp/queries'
import getTasks from '@wasp/queries/getTasks'

const MainPage = () => {
  const {
    data: allTasks,
    error: error1
  } = useQuery(fetchAllTasks)

  const {
    data: doneTasks,
    error: error2

>   } = useQuery(fetchFilteredTasks, { isDone: true})

Author: This addition fixes #329.

To prevent information leakage, the server won't forward these fields for any other HTTP status codes.

#### Using Entities in Queries
In most cases, resources used in Queries will be [Entities](#entity).
To use an Entity in your Query, add it to the declaration in Wasp:

```c {4,9} title="main.wasp"
// ...
query fetchAllTasks {
  fn: import { getAllTasks } from "@ext/queries.js",
  entities: [Task]
}

query fetchFilteredTasks {
  fn: import { getFilteredTasks } from "@ext/queries.js",
  entities: [Task]
}
```

Wasp will inject the specified Entity into the Query's context argument, giving you access to the Entity's Prisma API:
```js title="ext/queries.js"
// ...

export const getAllTasks = async (args, context) => {
  return context.entities.Task.findMany({})
}

export const getFilteredTasks = async (args, context) => {
  return context.entities.Task.findMany({ 
    where: { isDone: args.isDone }
  })
}
```

The object `context.entities.Task` exposes `prisma.task` from [Prisma's CRUD API](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/crud).

> ### Action

Author: Since the "Cache Invalidation" section refers to both Actions and Queries, I flipped the ordering to provide the necessary context.

However, Wasp will generate the code that lets you call the Query from any point in your code (i.e., client, server) using the same interface.
In other words, you won't have to worry about building an HTTP API for the Query, handling the request on the server, or even handling and caching the responses on the client.
Instead, simply focus on the business logic inside your Query and let Wasp take care of the rest!

To create a Wasp Query, you must:
1. Define the Query's NodeJS implementation
2. Declare the Query in Wasp using the `query` declaration

After completing these two steps, you'll be able to use the Query from any point in your code. 

#### Defining the Query's NodeJS implementation
A Query must be implemented as an `async` NodeJS function that takes two arguments.
Since both arguments are positional, you can name the parameters however you want, but we'll stick with `args` and `context`:
1. `args`:  An object containing all the arguments (i.e., payload) sent to the Query (e.g., filtering conditions).
Take a look at [the examples of usage](#using-the-query) to see how to pass this object to the Query.

> 3. `context`: An additional context object. This object contains user session information, as well as information about entities. The examples here won't use the context for simplicity purposes. You can read more about it in the [section about using entities in queries](#using-entities-in-queries).

Reviewer 3: Should we emphasize here that `args` are given on the call site, so basically it is whatever you pass to the Query, while `context` is provided by Wasp?
I am not sure if this is the right way to put it, I guess I am trying to say `args` are given by dev while `context` is wasp stuff.

Author: Definitely, I'll find a way to make it clear.

Here's an example of two simple Queries:
```js title="ext/queries.js"
// our "database"
const tasks = [
  { id: 1, description: "Buy some eggs", isDone: true },
  { id: 2, description: "Make an omelette", isDone: false },
  { id: 3, description: "Eat breakfast", isDone: false }
]

// You don't need to use the arguments if you don't need them
export const getAllTasks = async () => {
  return tasks;
}

> // The 'args' object is something sent by the caller (most often from the client) 

Reviewer 3: Aha nice I see now you mentioned it here that args are coming from the caller! I would still maybe mention it up there, at least a bit.

export const getFilteredTasks = async (args) => {
  const { isDone } = args;
  return tasks.filter(task => task.isDone === isDone)
}
```

#### Declaring a Query in Wasp 
After implementing your Queries in NodeJS, all that's left to do before using them is tell Wasp about it!
You can easily do this with the `query` declaration, which supports the following fields:

> - `fn: ExtImport` (required) - The import statement of the Query's NodeJs implementation 
> - `entities: [Entity]` (optional) - A list of entities you wish to use inside your Query

Reviewer 3: ```suggestion
- `fn: ExtImport` (required) - The import statement of the Query's NodeJs implementation. 
- `entities: [Entity]` (optional) - A list of entities you wish to use inside your Query.
```

import fetchFilteredTasks from '@wasp/queries/fetchFilteredTasks.js'

// ...

const allTasks = await fetchAllTasks();
const doneTasks = await fetchFilteredTasks({isDone: true})
```

**NOTE**: Wasp will not stop you from importing a Query's NodeJS implementation from `./queries.js` and calling it directly. However, we advise against this, as you'll lose all the useful features a Wasp Query provides (e.g., entity injection).

#### The `useQuery` hook
When using Queries on the client, you can make them reactive with the help of the `useQuery` hook.
This hook comes bundled with Wasp and is a thin wrapper around the `useQuery` hook from [_react-query_](https://github.com/tannerlinsley/react-query).

Wasp's `useQuery` hook takes three arguments:
- `queryFn` (required): A Wasp query declared in the previous step or, in other words, the client-side query function generated by Wasp based on a `query` declaration.
- `queryFnArgs` (optional): The arguments object (payload) you wish to pass into the query. The query's NodeJS implementation will receive this object as its first positional argument.
- `config` (optional): A _react-query_ `config` object.

> Wasp's `useQuery` hook behaves mostly the same as [_react-query_'s `useQuery` hook](https://react-query.tanstack.com/docs/api#usequery), the only difference being in not having to supply the key. Wasp handles the caching automatically under the hood.

Reviewer 3: I would say that Wasp handles not caching, but assigning a key, automatically under the hood.

Here's an example of calling the Queries using the `useQuery` hook:
```jsx
import React from 'react'
import { useQuery } from '@wasp/queries'

import fetchAllTasks from '@wasp/queries/fetchAllTasks'
import fetchFilteredTasks from '@wasp/queries/fetchFilteredTasks'

const MainPage = () => {
  const {
    data: allTasks,
    error: error1
  } = useQuery(fetchAllTasks)

  const {
    data: doneTasks,
    error: error2

>   } = useQuery(fetchFilteredTasks, { isDone: true})

Reviewer 3: ```suggestion
  } = useQuery(fetchFilteredTasks, { isDone: true })
```

  return (
    <div> 
        <p>All tasks: { JSON.stringify(allTasks || error1) }</p>
        <p>Finished tasks: { JSON.stringify(doneTasks || error2) }</p>
    </div>
  )
}

export default MainPage
```

#### Error Handling
For security reasons, all exceptions thrown in the Query's NodeJS implementation are sent to the client as responses with the HTTP status code `500`, with all other details removed.
Removing error information helps against accidentally leaking possibly sensitive information over the network.

If you do want to pass additional error information to the client, you can construct and throw an appropriate `HttpError` in your NodeJS Query function:
```js title=ext/queries.js
import HttpError from '@wasp/core/HttpError.js'

export const getTasksError = async (args, context) => {
  const statusCode = 403
  const message = 'You can\'t do this!'
  const data = { foo: 'bar' }
  throw new HttpError(statusCode, message, data)
}
```

> If the status code is `4xx`, the client will receive a response object with the corresponding `.message` and `.data` fields.

Reviewer 3: This will actually get rethrown as an Error on the client -> that was mentioned before but now it is not anymore.

To prevent information leakage, the server won't forward these fields for any other HTTP status codes.

#### Using Entities in Queries
In most cases, resources used in Queries will be [Entities](#entity).
To use an Entity in your Query, add it to the declaration in Wasp:

```c {4,9} title="main.wasp"
// ...
query fetchAllTasks {
  fn: import { getAllTasks } from "@ext/queries.js",
  entities: [Task]
}

query fetchFilteredTasks {
  fn: import { getFilteredTasks } from "@ext/queries.js",
  entities: [Task]
}
```

> Wasp will inject the specified Entity into the Query's context argument, giving you access to the Entity's Prisma API:

Reviewer 3: ```suggestion
Wasp will inject the specified Entity into the Query's `context` argument, giving you access to the Entity's Prisma API:
```

  })
}
```

The object `context.entities.Task` exposes `prisma.task` from [Prisma's CRUD API](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/crud).

### Action

Actions are very similar to Queries. So similar, in fact, we will only list the differences:
1. They can (and most often should) modify state, while Queries are only allowed to read it.
2. Since Actions don't need to be reactive, Wasp doesn't provide a React hook for them (like `useQuery` for Queries) - you just call them directly.
3. `action` declarations in Wasp are mostly identical to `query` declarations. The only difference is in the declaration's name. 

Here's an implementation of a simple Action:

```js title=actions.js
export const sayHi = async () => {
  console.log('The client said Hi!')
}
```
Its corresponding declaration in Wasp:

```c title="main.wasp"
// ...

action sayHi {
  fn: import { sayHi } from "@ext/actions.js"
}
```
And an example of how to import and call the declared Action:

```js
import sayHi from '@wasp/actions/sayHi'

// ...

sayHi()
```

More differences and Action/Query specific features will come in future versions of Wasp.

### Cache Invalidation

> One of the trickiest parts of managing a web app's state is making sure the data shown by the queries are up to date.

Author: I agree that the singular is more clear, I think this was Grammarly's suggestion. However, if we're both prefer the singular, I think it makes sense to take a little poetic license :).

It's possible to invalidate the caches manually through several mechanisms _react-query_ provides (e.g., refetch, direct invalidation).
However, since manual cache invalidation quickly becomes complex and error-prone, Wasp offers a quicker and a more effective solution to get you started: **automatic Entity-based Query cache invalidation**.
Because Actions can (and most often do) modify the state while Queries read it, Wasp invalidates a Query's cache whenever an Action that uses the same Entity is executed.

For example, let's assume that Action `A1` and Query `Q1` both use Entity `E1`. If `A1` is executed, `Q1`'s cached result may no longer be up-to-date.
Wasp will therefore invalidate it, making `Q1` up to date again.

In practice, this means that Wasp keeps the queries "fresh" without requiring you to think about cache invalidation.

On the other hand, such excessive cache invalidation can quickly become wasteful and will not work for any resource other than Entities.
Wasp does not yet provide native support for such use cases, so please use _react-query_'s built-in mechanisms for the time being.

> You can expect Wasp to support a more elegant solution in the future.

Reviewer 3: Maybe for this part I would stick with the original
```
On the other hand, this kind of automatic invalidation of cache can be wasteful (updating when not needed) and will not work if other resources than entities are used. In that case, make sure to use mechanisms provided by react-query for now, and expect more direct support in Wasp for handling those use cases in a nice, elegant way.
```

statement, as it is a bit more optimistic / exciting (`nice, elegant solution` vs `more elegant solution`).
